### DISCUSSION TITLE
Question about Usage

### DISCUSSION THREAD

**Discussion by kuro337** [bytedance/sonic#479] [Q&A] [answered]
Hello! I am trying to figure out capabilities of this library - and whether it can parse into a Struct without knowing the shape. 

Also I am trying to find out how to print a proper indented version from an API Response. 

For example so far I have this - 

```go
func MakeAPICallSonic(url string) error {
	resp, err := http.Get(url)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return err
	}
	
	// Is this recommended way of initializing it? Trying to understand when to use AST versus Decoder
	root, err := sonic.Get(body)
	if err != nil {
		return err
	}

	// PRINT RAW JSON using Sonic
	rawJSON, _ := root.Raw()
	fmt.Printf("JSON Response:\n%s\n", rawJSON)

	// PRINT RAW JSON using Sonic - returns result as a string similar to encoding/json
	jsonResp, _ := root.MarshalJSON()
	fmt.Printf("JSON Response:\n%s\n", jsonResp)


	return nil
}
```

Couple Questions - 
1. If I make a call to an REST API that returns a body - is ` ioutil.ReadAll(resp.Body)` required or can the resp.Body be passed directly using Sonic? 

2. Pretty Print JSON
Wondering how to do something like this - 
```go
PrettyPrintJSON("https://api.sampleapis.com/beers/ale")

func PrettyPrintJSON(url string) error {
	resp, err := http.Get(url)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return err
	}

	var result interface{}
	err = json.Unmarshal(body, &result)
	if err != nil {
		return err
	}

	prettyJSON, err := json.MarshalIndent(result, "", "  ")
	if err != nil {
		return err
	}

	fmt.Printf("Pretty Printed JSON\n\n%s", prettyJSON)

	return nil
}
```

3. Can Sonic parse a byte array or string representing the JSON response into an `interface{}` to get a Struct that can be accessed by the fields of the response? 

4. Difference between use cases between Encoder/Decoder and AST. 

Sorry for several questions - understanding idiomatic ways to use the Library based on how it has been designed. Looking to incorporate usage incrementally into large hyper scale distributed real time systems - and want to understand API design to build on top of. ``

Would really appreciate any help - amazing library and looking forward to using it more!

**Comment by AsterDY** [comment, marked as answer]
For question 1: It's mainly about cocurrency safty --- you can pass any `[]byte` to sonic as long as you can ensure it is not being used by other goroutines at the same time.
For question 2: At present sonic doesn't implement json formalization by itself. You can use `encoding/json`'s API to print
 For question 3: not totally understand your question. I guess you mean you didn't know the whole schema for json but has some  known fields --- for this situation you have two choice: 
- sonic supports partial schema and skip unknown fields, therefore you can define a partial struct only having known fields to be used as argument of `Unmarshal()`
- use `sonic.Get()`. Translate these known fields into pathes (the secondary arguments of `Get()`), and call `Get()` to get json AST nodes. After that. you can use `ast.Node.Interface()` to convert to corresponding go primitive values or use `ast.Node.Raw()` to get raw json message
